src/main/scala/is/hail/io/bgen/BgenWriter.scala

```diff
@@ -0,0 +1,290 @@
+package is.hail.io.bgen
+
+import is.hail.annotations._
+import is.hail.utils._
+import is.hail.variant.{Genotype, Variant, VariantDataset}
+
+object BgenWriter {
+  val ploidy: Byte = 2
+  val phased: Byte = 0
+
+  def shortToBytesLE(bb: ArrayBuilder[Byte], i: Int) {
+    bb += (i & 0xff).toByte
+    bb += ((i >>> 8) & 0xff).toByte
+  }
+
+  def intToBytesLE(bb: ArrayBuilder[Byte], i: Int) {
+    bb += (i & 0xff).toByte
+    bb += ((i >>> 8) & 0xff).toByte
+    bb += ((i >>> 16) & 0xff).toByte
+    bb += ((i >>> 24) & 0xff).toByte
+  }
+
+  def updateIntToBytesLE(bb: ArrayBuilder[Byte], i: Int, pos: Int) {
+    bb(pos) = (i & 0xff).toByte
+    bb(pos + 1) = ((i >>> 8) & 0xff).toByte
+    bb(pos + 2) = ((i >>> 16) & 0xff).toByte
+    bb(pos + 3) = ((i >>> 24) & 0xff).toByte
+  }
+
+  def stringToBytesWShortLength(bb: ArrayBuilder[Byte], s: String): Int = {
+    val bytes = s.getBytes
+    val l = bytes.length
+    shortToBytesLE(bb, l)
+    bb ++= bytes
+    2 + l
+  }
+
+  def stringToBytesWIntLength(bb: ArrayBuilder[Byte], s: String): Int = {
+    val bytes = s.getBytes
+    val l = bytes.length
+    intToBytesLE(bb, l)
+    bb ++= bytes
+    4 + l
+  }
+
+  def bitPack(bb: ArrayBuilder[Byte], input: Array[Int], nBitsPerProb: Int) {
+    require(nBitsPerProb >= 1 && nBitsPerProb <= 32, "Number of bits per probability must be between 1 and 32 inclusive.")
+
+    val bitMask = (1L << nBitsPerProb) - 1
+    var byteIndex = 0
+    var data = 0L
+    var dataSize = 0
+
+    var i = 0
+    while (i < input.length) {
+      data |= ((input(i).toUIntFromRep.toLong & bitMask) << dataSize)
+      dataSize += nBitsPerProb
+
+      while (dataSize >= 8) {
+        bb += data.toByte
+        data = data >>> 8
+        dataSize -= 8
+        byteIndex += 1
+      }
+      i += 1
+    }
+
+    if (dataSize > 0)
+      bb += data.toByte
+  }
+
+  def sortedIndex(a: Array[Double], idx: Array[Int]) {
+    val n = a.length
+    if (n == 3) {
+      if (a(0) >= a(1) && a(1) >= a(2)) {
+        idx(0) = 0
+        idx(1) = 1
+        idx(2) = 2
+      } else if (a(0) >= a(2) && a(2) >= a(1)) {
+        idx(0) = 0
+        idx(1) = 2
+        idx(2) = 1
+      } else if (a(1) >= a(0) && a(0) >= a(2)) {
+        idx(0) = 1
+        idx(1) = 0
+        idx(2) = 2
+      } else if (a(1) >= a(2) && a(2) >= a(0)) {
+        idx(0) = 2
+        idx(1) = 0
+        idx(2) = 1
+      } else if (a(2) >= a(0) && a(0) >= a(1)) {
+        idx(0) = 1
+        idx(1) = 2
+        idx(2) = 0
+      } else {
+        idx(0) = 2
+        idx(1) = 1
+        idx(2) = 0
+      }
+    } else {
+      throw new UnsupportedOperationException("sortedIndex is only supported for arrays of length 3.")
```

Comment:
I prefer to to put checks like this at the top.  `if (a.length != 3) throw Blah("foo") ...`.